doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,329 @@
+other modules.
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Get a reusable channel object
+const channel = diagnostics_channel.channel('my-channel');
+
+// Subscribe to the channel
+channel.subscribe((message, name) => {
+  // Received data
+});
+
+// Check if the channel has an active subscriber
+if (channel.hasSubscribers) {
+  // Publish data to the channel
+  channel.publish({
+    some: 'data'
+  });
+}
+
+// Spans correlate a sequence of events
+const span = channel.span({
+  some: 'start data'
+});
+
+// Annotations are for data not strictly related to the start or end events
+span.annotate({
+  some: 'annotation'
+});
+
+// This end event will share an `id` with the span started above
+span.end({
+  some: 'end data'
+});
+```
+
+#### `diagnostics_channel.hasSubscribers(name)`
+
+* `name` {string|symbol} The channel name
+* Returns: {boolean} If there are active subscribers
+
+Check if there are active subscribers to the named channel. This is helpful if
+the message you want to send might be expensive to prepare.
+
+This API is optional but helpful when trying to publish messages from very
+performance-senstive code.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+if (diagnostics_channel.hasSubscribers('my-channel')) {
+  // There are subscribers, prepare and publish message
+}
+```
+
+#### `diagnostics_channel.channel(name)`
+
+* `name` {string|symbol} The channel name
+* Returns: {Channel} The named channel object
+
+This is the primary entry-point for anyone wanting to interact with a named
+channel. It produces a channel object which is optimized to reduce overhead at
+publish time as much as possible.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+```
+
+### Class: `Channel`
+
+The class `Channel` represents an individual named channel within the data
+pipeline. It is use to track subscribers and to publish messages when there
+are subscribers present. It exists as a separate object to avoid channel
+lookups at publish time, enabling very fast publish speeds and allowing
+for heavy use while incurring very minimal cost. Channels are created with
+[`diagnostics_channel.channel(name)`][], constructing a channel directly
+with `new Channel(name)` is not supported.
+
+#### `channel.hasSubscribers`
+
+* Returns: {boolean} If there are active subscribers
+
+Check if there are active subscribers to this channel. This is helpful if
+the message you want to send might be expensive to prepare.
+
+This API is optional but helpful when trying to publish messages from very
+performance-senstive code.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+if (channel.hasSubscribers) {
+  // There are subscribers, prepare and publish message
+}
```

Comment:
I'm still generally concerned about this pattern -- I understand what it's looking to accomplish, but it still concerns me... I'd much rather remove the `channel.hasSubscribers` and just rely in `channel.publish()` (that is, just attempt to do the thing without the gates).

By allowing `message` to be either a function or an object with a special symbol property, we can defer potentially expensive work until after the internal check that publish does:

```js
const { channel, diagnosticsInfoSymbol } = require('diagnostics_channel');
channel.publish({
  [diagnosticsInfoSymbol]() {
    let message;
    // do the expensive diagnostic computation here
    return message;
  }
});
```

This avoids the potential footgun where the subscribers change in between the initial call to `hasSubscribers` and the actual publish -- and since publish is likely needing to have it's own internal check for subscribers, saves us one check

Reply:
The `hasSubscribers` check is entirely optional. If you just call `publish(...)` without a check and there's nothing to publish to, it'll do nothing. The `hasSubscribers` check is purely for very hot code paths where an object allocation might actually have significant implications. Using `hasSubscribers` should _almost never_ be needed. It's mainly there just so core can use it as-needed. I'm fine with not documenting it, if you think that would be better.

As for the suggestion of a function as the message or as a special symbol property, that creates an extra closure which is likely much worse for the performance than anything the `hasSubscribers` check would be trying to cover for. Also, a closure introduces the chance of holding memory longer than intended, and possibly raising things from being stack allocated to heap allocated. Given the performance-sensitive nature of this API, I'd strongly recommend against a design like that.